File: Generated_Tutorials/cpp_tutorial/06_LinkedLists/solutions/reverse_list.cpp
#include <iostream>

struct Node {
    int data;
    Node* next;
    Node(int val) : data(val), next(nullptr) {}
};

void printList(Node* head) {
    while (head) {
        std::cout << head->data << " -> ";
        head = head->next;
    }
    std::cout << "NULL\n";
}

Node* reverseList(Node* head) {
    Node* prev = nullptr;
    Node* current = head;
    Node* next = nullptr;

    while (current != nullptr) {
        next = current->next; // Store next
        current->next = prev; // Reverse
        prev = current;       // Move prev
        current = next;       // Move current
    }
    return prev; // Prev is the new head
}

int main() {
    Node* head = new Node(1);
    head->next = new Node(2);
    head->next->next = new Node(3);
    head->next->next->next = new Node(4);

    std::cout << "Original List: ";
    printList(head);

    head = reverseList(head);

    std::cout << "Reversed List: ";
    printList(head);

    return 0;
}

File: Generated_Tutorials/java_tutorial/06_LinkedLists/exercises/ReverseList.java
class Node {
    int data;
    Node next;
    Node(int d) { data = d; }
}

public class ReverseList {

    static void printList(Node node) {
        while (node != null) {
            System.out.print(node.data + " -> ");
            node = node.next;
        }
        System.out.println("NULL");
    }

    // TODO: Reverse the linked list
    static Node reverse(Node head) {
        /*
        Node prev = null;
        Node current = head;
        Node next = null;
        
        while (current != null) {
            // Logic here
        }
        return prev;
        */
        return null; // Replace
    }

    public static void main(String[] args) {
        Node head = new Node(1);
        head.next = new Node(2);
        head.next.next = new Node(3);

        System.out.print("Original: ");
        printList(head);

        // head = reverse(head);

        System.out.print("Reversed: ");
        printList(head);
    }
}
